File: 3217.delete-nodes-from-linked-list-present-in-array.py
#
# @lc app=leetcode id=3217 lang=python3
#
# [3217] Delete Nodes From Linked List Present in Array
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def modifiedList(self, nums: List[int], head: Optional[ListNode]) -> Optional[ListNode]:
        # Convert nums to a set for O(1) lookups
        num_set = set(nums)
        
        # Use a dummy node to handle head removal easily
        dummy = ListNode(0)
        dummy.next = head
        
        curr = dummy
        while curr.next:
            if curr.next.val in num_set:
                # Skip the next node
                curr.next = curr.next.next
            else:
                # Move to the next node
                curr = curr.next
                
        return dummy.next
# @lc code=end

File: 3217.delete-nodes-from-linked-list-present-in-array.cpp
#
# @lc app=leetcode id=3217 lang=cpp
#
# [3217] Delete Nodes From Linked List Present in Array
#
# @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* modifiedList(vector<int>& nums, ListNode* head) {
        // Create a set for O(1) lookup
        unordered_set<int> numSet(nums.begin(), nums.end());
        
        // Create a dummy node to simplify edge cases
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        
        // Traverse the list
        ListNode* prev = dummy;
        ListNode* curr = head;
        
        while (curr != nullptr) {
            if (numSet.count(curr->val)) {
                // Current node should be deleted
                prev->next = curr->next;
            } else {
                // Move prev forward
                prev = curr;
            }
            curr = curr->next;
        }
        
        return dummy->next;
    }
};
# @lc code=end